### ISSUE TITLE
Change Menu behavior

### ISSUE THREAD

**Issue by GreyEcologist** [CosmicMind/Material#595]
the Menu behavior has changed since 2.x, I want to keep using it, but I don't need the MenuViewController. I've looked in the sample projects but couldn't find an example of how to use Menu directly.

```
        let menu = Menu(frame: CGRect(x: 0, y: 0, width: 50, height: 50))

        let btn1: FabButton = FabButton()
        btn1.depthPreset = .depth2
        btn1.pulseAnimation = .centerWithBacking
        btn1.backgroundColor = buttonColor
        btn1.tintColor = Color.white
        btn1.setImage(R.image.ic_add_white()!.withRenderingMode(.alwaysTemplate), for: .normal)
        btn1.setImage(R.image.ic_add_white()!.withRenderingMode(.alwaysTemplate), for: .highlighted)
        btn1.addTarget(self, action: #selector(self.handleMenu(_:)), for: .touchUpInside)
        menu.addSubview(btn1)
        btn1.position = origin
        
        var image = R.image.facebooklogo()!
        let btn2: FabButton = FabButton()
        btn2.tag = 1
        btn2.clipsToBounds = true
        btn2.contentMode = .scaleToFill
        btn2.depthPreset = .depth1
        btn2.pulseColor = Color.white
        btn2.backgroundColor = Color.indigo.darken1
        btn2.borderColor = Color.white
        btn2.borderWidth = 2
        btn2.setImage(image, for: .normal)
        btn2.setImage(image, for: .highlighted)
        btn2.addTarget(self, action: #selector(self.socialButton(_:)), for: .touchUpInside)
        menu.addSubview(btn2)
        
        image = R.image.twitterlogo()!
        let btn3: FabButton = FabButton()
        btn3.tag = 2
        btn3.contentMode = .scaleToFill
        btn3.clipsToBounds = true
        btn3.depthPreset = .depth1
        btn3.backgroundColor = Color.lightBlue.base
        btn3.pulseColor = Color.white
        btn3.borderColor = Color.white
        btn3.borderWidth = 2
        btn3.setImage(image, for: .normal)
        btn3.setImage(image, for: .highlighted)
        btn3.addTarget(self, action: #selector(self.socialButton(_:)), for: .touchUpInside)
        menu.addSubview(btn3)
        
        menu.direction = .up
        menu.baseSize = CGSize(width: 50, height: 50)
        menu.delegate = self
        menu.views = [btn1, btn2, btn3]

       self.view.addSubview(menu)
```

I've tried to update my implementation to match the changes but the Menu doesn't respond when clicked on it and the FabButton don't display anymore.

**Comment by danieldahan** [comment]
Hey, yes, the `MenuView` and `Menu` became one in Material 2, so now `Menu` is effectively the old `MenuView` with the mechanics of the previous `Menu`. 

So  few things I am seeing here. 

1. You can omit these lines: `menu.addSubview(btn3)`, when you set `menu.views = [btn1, btn2, btn3]` it automatically adds the views to the menu as a subview. 

2. These lines can be omitted as well:

```swift
btn3.setImage(image, for: .normal)
btn3.setImage(image, for: .highlighted)
```

and you can now use this one line:

```swift
btn3.image = image
```

3. This line looks off: 

```swift
Menu(frame: CGRect(x: 0, y: 0, width: 50, height: 50))
```

It looks like you are adding the menu to the top left corner, is this correct? And, you need to ensure that it is not behind anything and is brought to the front. 

Try these updates, and let me know, I would be happy to help further :)

**Comment by GreyEcologist** [comment]
Hey Daniel, after some reviews it looks like the new Menu works fine, the issue I'm running into is after I apply the animations below, the frame of the Menu is not updated:

```
    func showingButtons() {
        let first: Button? = self.socialButton.views.first as? Button
        first!.animate(animation: Motion.animate(group: [
            Motion.rotate(rotation: 3),
            Motion.translationX(by: 120)
            ]))
        
        self.voteButton.animate(animation: Motion.animate(group: [
            Motion.rotate(rotation: 3),
            Motion.translationX(by: -120)
            ]))
        
        for subview in self.socialButton.views.enumerated() {
               subview.element.frame.offsetBy(dx: 120, dy: 0)
        }
    }
```

The socialButton is a Menu, the voteButton is a Button but they both have the same issue. If I remove the animation then all works fine, but with animation, after the animate() has completed, the frame of both remains in their original positions. 

Which is weird because I can see the buttons in the simulators but nothing happens when I click on them (using a breakpoint and description shows the frame is never updated).

I've also tried animating the Menu directly rather than the subviews:

```
        self.socialButton.animate(animation: Motion.animate(group: [
            Motion.rotate(rotation: 3),
            Motion.translationX(by: Constants.kPadding + 100)
            ]))
```

Same results. It seems like somehow the animation updates the view but not the frame?

**Comment by danieldahan** [comment]
You should be animation the Menu in this case based on how you have it setup. Is it possible to share a sample with me and I can help fix it up?

**Comment by GreyEcologist** [comment]
https://github.com/GreyEcologist/MaterialButtons

If you uncomment the lines inside the showingButtons() you can see the non-responsiveness.

**Comment by danieldahan** [comment]
Hey, so I looked at your sample and I noticed a lot of fighting between layout, frame, and animation. So I wrote a simple sample for you to replace your current code with:

```swift
import UIKit
import Material

class ViewController: UIViewController {
    
    fileprivate let menu: Menu
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        prepareMenu()
        
        view.layout(menu).size(menu.baseSize).bottom(24).right(24)
    }

    required init?(coder aDecoder: NSCoder) {
        menu = Menu()
        super.init(coder: aDecoder)
    }
    
    override func viewDidAppear(_ animated: Bool) {
        super.viewDidAppear(animated)
        showingButtons()
    }
    
    func showingButtons() {
        menu.views.first?.animate(animation: Motion.animate(group: [
            Motion.rotate(rotation: 3),
            Motion.translationY(by: -120)
        ]))
    }
    
    func handleMenu(_ open: Bool = false) {
        print("handled")
    }
    
    func helpAction() {
        print("helpAction")
    }
    
    func prepareMenu() {
        let btn1 = FabButton()
        btn1.depthPreset = .depth2
        btn1.pulseAnimation = .centerWithBacking
        btn1.backgroundColor = Color.red
        btn1.tintColor = Color.white
        btn1.addTarget(self, action: #selector(handleMenu(_:)), for: .touchUpInside)
        
        let btn2 = FabButton()
        btn2.tag = 1
        btn2.clipsToBounds = true
        btn2.contentMode = .scaleToFill
        btn2.depthPreset = .depth1
        btn2.pulseColor = Color.white
        btn2.backgroundColor = Color.indigo.darken1
        btn2.borderColor = Color.white
        btn2.borderWidth = 2
        
        let btn3 = FabButton()
        btn3.tag = 2
        btn3.contentMode = .scaleToFill
        btn3.clipsToBounds = true
        btn3.depthPreset = .depth1
        btn3.backgroundColor = Color.lightBlue.base
        btn3.pulseColor = Color.white
        btn3.borderColor = Color.white
        btn3.borderWidth = 2
        
        menu.direction = .up
        menu.views = [btn1, btn2, btn3]
    }
}
```

Look at this code and see how things were setup. This should help you get going in the right direction :)

**Comment by GreyEcologist** [comment]
@danieldahan Hey thanks for the layout tips, that make things a lot cleaner.

However, when I click on the button after the animation it still doesn't respond, but I click on its former position, where the button was before the animation, then I get a response. I'm using Material 2.3.10.

**Comment by danieldahan** [comment]
So I figure this out. It turns out that iOS has changed the animationDidStop method which updates the final values to a delegation call `CAAnimationDelegate`. So I made a fix for this and will push a new release :)

**Comment by danieldahan** [comment]
Fixed and released in [Material 2.3.12](https://github.com/CosmicMind/Material/releases/tag/2.3.12).

**Comment by GreyEcologist** [comment]
Great! Thanks Dan!
